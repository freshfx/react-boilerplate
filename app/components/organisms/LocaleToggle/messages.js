/*
 * LocaleToggle Messages
 *
 * This contains all the text for the LanguageToggle component.
 */
import {defineMessages} from 'react-intl'

export default defineMessages({
  de: {
    defaultMessage: 'de',
    id: 'boilerplate.containers.LocaleToggle.de'
  },
  en: {
    defaultMessage: 'en',
    id: 'boilerplate.containers.LocaleToggle.en'
  }
})
