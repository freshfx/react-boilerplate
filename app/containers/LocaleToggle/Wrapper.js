import styled from '@emotion/styled'

const Wrapper = styled.div`
  padding: 2px;
`

export default Wrapper
