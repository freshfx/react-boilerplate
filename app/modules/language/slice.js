import {createSlice} from '@reduxjs/toolkit'
import deepFreeze from 'deep-freeze'

import {DEFAULT_LOCALE} from '../../i18n'

const initialState = deepFreeze({
  locale: DEFAULT_LOCALE
})

const name = 'language'
const {actions, reducer} = createSlice({
  initialState,
  name,
  reducers: {
    changeLocale: (state, action) => {
      const {locale} = action.payload
      state.locale = locale
    }
  }
})

export {actions, initialState, name, reducer}

export default {
  key: name,
  reducer
}
