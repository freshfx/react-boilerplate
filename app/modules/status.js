/* eslint-disable sort-keys */
const STATUS = {
  IDLE: 'idle',
  PENDING: 'pending',
  SUCCESS: 'success',
  FAILURE: 'failure'
}
/* eslint-enables */

const AVAILABLE_STATUS = Object.values(STATUS)

export {AVAILABLE_STATUS}
export default STATUS
