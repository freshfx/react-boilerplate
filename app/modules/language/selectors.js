import {
  createSliceFieldSelector,
  createSliceSelector
} from 'modules/utils/common-selectors'

import {initialState, name} from './slice'

const selectSlice = createSliceSelector(name, initialState)
const selectLocale = createSliceFieldSelector(selectSlice, 'locale')

export {selectLocale, selectSlice}
