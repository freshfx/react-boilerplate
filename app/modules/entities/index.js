import reducer, {actions} from './slice'
import * as selectors from './selectors'

export {
  actions,
  selectors
}

export default reducer
