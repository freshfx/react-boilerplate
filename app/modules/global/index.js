/* eslint-disable import/prefer-default-export,import/no-unused-modules */
import * as selectors from './selectors'

export {selectors}
