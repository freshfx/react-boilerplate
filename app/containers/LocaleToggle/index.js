/*
 *
 * LanguageToggle
 *
 */

import React from 'react'
import {compose} from 'redux'
import {useDispatch, useSelector} from 'react-redux'

import {appLocales} from 'i18n'
import Toggle from 'components/Toggle'
import {actions, selectors} from 'modules/language'

import Wrapper from './Wrapper'
import messages from './messages'

const LocaleToggle = () => {
  const dispatch = useDispatch()
  const locale = useSelector(selectors.selectLocale)

  const onChangeLocale = ({target: {value}}) => {
    dispatch(actions.changeLocale({locale: value}))
  }

  return (
    <Wrapper>
      <Toggle
        value={locale}
        values={appLocales}
        messages={messages}
        onToggle={onChangeLocale}
      />
    </Wrapper>
  )
}

export default compose(React.memo)(LocaleToggle)
