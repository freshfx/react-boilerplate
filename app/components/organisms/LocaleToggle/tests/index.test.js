import React from 'react'
import {fireEvent, render} from '@testing-library/react'
import {IntlProvider} from 'react-intl'
import {Provider} from 'react-redux'

import configureStore from 'configure-store'
import {actions} from 'modules/language'
import history from 'utils/history'

import LocaleToggle from '../index'

const store = configureStore({}, history)
jest.spyOn(store, 'dispatch')

const renderComponent = (props = {}) =>
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <LocaleToggle {...props} />
      </IntlProvider>
    </Provider>
  )

describe('LocaleToggle', () => {
  beforeEach(() => {
    store.dispatch.mockClear()
  })

  it('should render the default language messages', () => {
    const {container} = renderComponent()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should change the locale on option click', () => {
    const {container} = renderComponent()
    const [select] = container.getElementsByTagName('select')
    const locale = 'de'
    fireEvent.change(select, {target: {value: locale}})
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(actions.changeLocale({locale}))
  })
})
