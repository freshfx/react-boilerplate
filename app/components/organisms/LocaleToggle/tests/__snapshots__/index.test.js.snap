// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocaleToggle should render the default language messages 1`] = `
.emotion-0 {
  padding: 2px;
}

.emotion-2 {
  line-height: 1em;
  background-color: transparent;
  border-style: none;
}

<div
  class="emotion-0 emotion-1"
>
  <select
    class="emotion-2 emotion-3"
  >
    <option
      value="en"
    >
      en
    </option>
    <option
      value="de"
    >
      de
    </option>
  </select>
</div>
`;
